posit-dev/positron
# (Start to) handle cli hyperlinks in the terminal used for pkg dev tasks

Addresses #5218

This PR makes suggested code hints like `testthat::snapshot_accept()` or `testthat::snapshot_review()` clickable and runnable from the integrated terminal where positron-r sends package development tasks, such as `devtools::test()`. "Runnable" in the sense that the code is run in the user's R console. This gets routed into the same handlers as clickable code entered directly in the console, so the existing safety guards are in place.

<img width="544" alt="Screenshot 2024-11-18 at 1 04 51 PM" src="https://github.com/user-attachments/assets/ddead5f3-ee5c-4f36-a50b-12f5ac7e6465">

(Note that this PR does _not_ change behaviour for file hyperlinks, which I've crossed out in the screenshot above. Those hyperlinks currently get delegated to the operating system, so for many R users, a `.R` file will open in RStudio. Changing that is a separate problem #5409.)

### QA Notes

To see the new behaviour, you need a development version of the cli R package. I would install that via `pak::pak("r-lib/cli")`. (Context: https://github.com/r-lib/cli/pull/739.)

Then you need to run tests or `R CMD check` on a package with a failing snapshot test, in order to get a clickable invitation to accept a new snapshot or to review the proposed snapshot in a Shiny app. I have made a toy package that could be obtained via [`usethis::create_from_github("jennybc/clilinks")`](https://github.com/jennybc/clilinks). It has a couple of snapshot tests that will always fail 😄 because they attempt to snapshot a random number.

* Install dev cli: `pak::pak("r-lib/cli")`
* Identify a package that tickles this feature and open it in Positron, perhaps via: `usethis::create_from_github("jennybc/clilinks")`
* Use our gesture for running package tests: Ctrl/Cmd + Shift + T or select *R: Test R Package* from the command palette.
* Click on a "runnable code" hyperlink:
  - Note this requires Ctrl/Cmd click! This is a VS Code convention, not specific to us.
  - Note that you might have to click twice. This is a bug (or 2 bugs?) for which fixes are making their way through the system (https://github.com/microsoft/vscode/issues/230010). Again, not us.
  - You will need to grant the extension permission to handle such URIs, either as a one-off or more persistently. This is a VS Code feature.

<img width="544" alt="Screenshot 2024-11-18 at 1 04 51 PM" src="https://github.com/user-attachments/assets/24a880e0-9ebe-4d88-8051-f859480b04fa">

<img width="285" alt="Screenshot 2024-11-18 at 12 55 37 PM" src="https://github.com/user-attachments/assets/d1187098-6a76-44b5-b5a5-be8752c43e5f">

If you click `testthat::snapshot_review()`, you should see the code execute in the R console and this Shiny app appears. Click on "Accept" (the fact that "Skip" seems broken is https://github.com/r-lib/testthat/issues/2025 and unrelated to this PR).

<img width="780" alt="Screenshot 2024-11-18 at 12 55 47 PM" src="https://github.com/user-attachments/assets/f7dff7ea-1bf4-4e52-8eb5-e92fd3481cda">
